servant-server/src/Servant/Server/Internal/RoutingApplication.hs

```diff
@@ -240,13 +248,19 @@ passToServer Delayed{..} x =
 runDelayed :: Delayed env a
            -> env
            -> Request
-           -> IO (RouteResult a)
-runDelayed Delayed{..} env = runDelayedIO $ do
-  c <- capturesD env
-  methodD
-  a <- authD
-  b <- bodyD
-  DelayedIO (\ req -> return $ serverD c a b req)
+           -> IO (RouteResult a, IO ())
+runDelayed Delayed{..} env req = do
+  cleanupRef <- newIORef (return ())
+  routeRes <- runDelayedIO
+    (do c <- capturesD env
+        methodD
+        a <- authD
+        b <- bodyD
+        liftIO (writeIORef cleanupRef $ cleanupD b)
```

Comment:
I would feel better if we were able to use something like `bracket` for this. Not sure that's easily doable.